polynomial/frozen.py

```diff
@@ -0,0 +1,96 @@
+"""This module defines the Freezable interface and subclasses."""
+from polynomial.core import Constant, Polynomial, extract_polynomial
+
+
+class Freezable:
+    """An interface for freezable objects."""
+
+    def _freeze(self):
+        """Prevent further modification of self."""
+        if not self._is_frozen():
+            self._frozen = True
+
+    def _is_frozen(self):
+        """Return true if self is frozen."""
+        return getattr(self, "_frozen", False)
+
+    def __setitem__(self, key, value):
+        """Implement self[x] = y; disallows setting item if frozen."""
+        if self._is_frozen():
+            raise AttributeError()
+        super().__setitem__(self, key, value)
+
+    def __setattr__(self, key, value):
+        """Implement self.x; disallows setting attr if frozen."""
+        if not self._is_frozen():
+            object.__setattr__(self, key, value)
+        else:
+            raise AttributeError("Can not modify frozen object.")
+
+
+class FrozenPolynomial(Freezable, Polynomial):
+    """A polynomial which can not be directly modified."""
+
+    def __init__(self, *args, **kwargs):
+        """Create a polynomial from the args, and then freeze it."""
+        Polynomial.__init__(self, *args, **kwargs)
+        self._trim = self._no_op
+        self._freeze()
```

Comment:
Why not making `self._vector` immutable as well (i.e. a `tuple`). This would fix an issue where things like `frozen_polynomial._vector[1] = 2` are possible and would allow for an implementation of `__hash__` (in order for the class to be fully immutable).